src/android/FileUtils.java

```diff
@@ -347,18 +348,25 @@ else if (action.equals("write")) {
             threadhelper( new FileOp( ){
                 public void run(JSONArray args) throws JSONException, FileNotFoundException, IOException, NoModificationAllowedException {
                     String fname=args.getString(0);
+                    String nativeURL = resolveLocalFileSystemURI(fname).getString("nativeURL");
                     String data=args.getString(1);
                     int offset=args.getInt(2);
                     Boolean isBinary=args.getBoolean(3);
-                    /*
-                     * If we don't have the package name in the path, we're reading and writing to places we need permission for
-                     */
-                    if(fname.contains(cordova.getActivity().getPackageName()) ||
-                            hasReadPermission()) {
-                        long fileSize = write(fname, data, offset, isBinary);
-                        callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, fileSize));
+
+                    JSONObject j = requestAllPaths();
+                    String[] allowedStorageDirectories = {j.getString("applicationStorageDirectory"), j.getString("externalApplicationStorageDirectory")};
+                    boolean needWritePermission = true;
+
+                    // Ask for the write permission if the native url lies outside the allowed storage directories
+                    for(String directory : allowedStorageDirectories) {
+                        if(nativeURL.startsWith(directory) || hasWritePermission()) {
```

Comment:
Is there a native API call (that does normalizing, etc.) to check if a path contains a path?